src/data_race.rs

```diff
@@ -0,0 +1,1189 @@
+    ///  this finds the two racing threads and the type
+    ///  of data-race that occured, this will also
+    ///  return info about the memory location the data-race
+    ///  occured in
+    #[cold]
+    #[inline(never)]
+    fn report_data_race<'tcx>(
+        global: &MemoryExtra, range: &MemoryCellClocks,
+        action: &str, is_atomic: bool,
+        pointer: Pointer<Tag>, len: Size
+    ) -> InterpResult<'tcx> {
+        let (current_index, current_clocks) = global.current_thread_state();
+        let write_clock;
+        let (
+            other_action, other_thread, other_clock
+        ) = if range.write > current_clocks.clock[range.write_index] {
+            // Convert the write action into the vector clock it
+            //  represents for diagnostic purposes
+            write_clock = VClock::new_with_index(range.write_index, range.write);
+            ("WRITE", range.write_index, &write_clock)
+        }else if let Some(idx) = Self::find_gt_index(
+            &range.read, &current_clocks.clock
+        ){
+            ("READ", idx, &range.read)
+        }else if !is_atomic {
+            if let Some(atomic) = range.atomic() {
+                if let Some(idx) = Self::find_gt_index(
+                    &atomic.write_vector, &current_clocks.clock
+                ) {
+                    ("ATOMIC_STORE", idx, &atomic.write_vector)
+                }else if let Some(idx) = Self::find_gt_index(
+                    &atomic.read_vector, &current_clocks.clock
+                ) {
+                    ("ATOMIC_LOAD", idx, &atomic.read_vector)
+                }else{
+                    unreachable!("Failed to find report data-race for non-atomic operation: no race found")
+                }
+            }else{
+                unreachable!("Failed to report data-race for non-atomic operation: no atomic component")
+            }
+        }else{
+            unreachable!("Failed to report data-race for atomic operation")
+        };
+
+        // Load elaborated thread information about the racing thread actions
+        let current_thread_info = global.print_thread_metadata(current_index);
+        let other_thread_info = global.print_thread_metadata(other_thread);
+        
+        // Throw the data-race detection
+        throw_ub_format!(
+            "Data race detected between {} on {} and {} on {}, memory({:?},offset={},size={})\
+            \n\t\t -current vector clock = {:?}\
+            \n\t\t -conflicting timestamp = {:?}",
+            action, current_thread_info, 
+            other_action, other_thread_info,
+            pointer.alloc_id, pointer.offset.bytes(), len.bytes(),
+            current_clocks.clock,
+            other_clock
+        )
+    }
+
+    /// Detect data-races for an unsychronized read operation, will not perform
+    ///  data-race threads if `multi-threaded` is false, either due to no threads
+    ///  being created or if it is temporarily disabled during a racy read or write
```

Comment:
s/racy/atomic? "racy" doesn't seem right.^^ We don't want to allow races, after all.